File: 145.binary_tree_postorder_traversal/binary_tree_postorder_traversal.cpp
class Solution
{
	public:
		vector<int> postorderTraversal(TreeNode *root)
		{
			vector<int> result;
			stack<const TreeNode *> s;
			const TreeNode *p = root, *pre = NULL;

			while(p != NULL || !s.empty())
			{
				while(p != NULL)
				{
					s.push(p);
					p = p->left;
				}
				p = s.top();
				if(p->right == NULL || p->right == pre)
				{
					result.push_back(p->val);
					s.pop();
					pre = p;
					p = NULL;
				}
				else
				{
					p = p->right;
				}
			}
			return result;
		}
};

File: 145.binary_tree_postorder_traversal/binary_tree_postorder_traversal.py
class Solution(object):
    def postorderTraversal(self, root):
        ret = []
        stack = []
        pre = None
        cur = root
        while cur or stack:
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack[-1]
            if cur.right == None or cur.right == pre:
                del stack[-1]
                ret.append(cur.val)
                pre = cur
                cur = None
            else:
                cur = cur.right
        return ret
